### PR TITLE
strategy:irr: a mean reversion based on box of klines in same direction

### PR DESCRIPTION
**Pull request by austin362667**
Due to the severe "delay" issue in binance 1s kline stream.
~~We've refactored irr strategy into backtesting mode (consume KLine onClose), and real-time mode (consume bookTicker which triggered every interval via goroutine)~~

~~Now we've redesigned the time ticker(goroutine) to consume aggerate trade market stream based on web socket .~~

Rollback to interval time trigger for a much easier programming model and really small intervals.

### CODE REVIEW

**Review by andycheng123 — APPROVED** [review]

File: pkg/strategy/irr/draw.go [hunk 1]
```diff
@@ -11,26 +11,43 @@ import (
...
 		var buffer bytes.Buffer
 		if err := canvas.Render(chart.PNG, &buffer); err != nil {
-			log.WithError(err).Errorf("cannot render cumpnl in drift")
-			reply.Message(fmt.Sprintf("[error] canot render cumpnl in drift: %v", err))
+			log.WithError(err).Errorf("cannot render nav in irr")
+			reply.Message(fmt.Sprintf("[error] canot render nav in irr: %v", err))
+			return
+		}
+		bbgo.SendPhoto(&buffer)
+		return
```
Comment (c9s) [inline pkg/strategy/irr/draw.go:37]: fix the lint issue

File: pkg/strategy/irr/strategy.go [hunk 2]
```diff
@@ -3,23 +3,25 @@ package irr
...
 	"github.com/sirupsen/logrus"
+	"go.uber.org/atomic"
 )
 
 const ID = "irr"
 
 var one = fixedpoint.One
 var zero = fixedpoint.Zero
-var Fee = 0.0008 // taker fee % * 2, for upper bound
+var Fee = 0.000 // taker fee % * 2, for upper bound
```
Comment (c9s) [inline pkg/strategy/irr/strategy.go:24]: make it private?

File: pkg/strategy/irr/strategy.go [hunk 3]
```diff
@@ -319,63 +345,124 @@ func (s *Strategy) Run(ctx context.Context, orderExecutor bbgo.OrderExecutor, se
...
+			s.obBuyPrice = atomic.NewFloat64(bt.Buy.Float64())
+			s.obSellPrice = atomic.NewFloat64(bt.Sell.Float64())
+		})
 
-	}))
+		s.session.MarketDataStream.OnAggTrade(func(trade types.Trade) {
+			s.currentTradePrice = atomic.NewFloat64(trade.Price.Float64())
+		})
+
+		go func() {
```
Comment (c9s) [inline pkg/strategy/irr/strategy.go:391]: this part is not backtest-able

File: pkg/strategy/irr/strategy.go [hunk 4]
```diff
@@ -398,3 +486,31 @@ func (s *Strategy) CalcAssetValue(price fixedpoint.Value) fixedpoint.Value {
...
+
+func (s *Strategy) rebalancePosition(bestBid, bestAsk float64, w float64) {
+	if w < 0.5 {
+		_, errB := s.orderExecutor.SubmitOrders(context.Background(), types.SubmitOrder{
+			Symbol:   s.Symbol,
+			Side:     types.SideTypeBuy,
+			Quantity: s.Quantity,
+			Type:     types.OrderTypeLimitMaker,
+			Price:    fixedpoint.NewFromFloat(bestBid - s.Spread),
+			Tag:      "irr short: buy",
```
Comment (c9s) [inline pkg/strategy/irr/strategy.go:498]: tag format is like this `irrShortBuy`

Comment (c9s) [reply pkg/strategy/irr/strategy.go:498]: but I guess you mean irrShortClose here

File: pkg/strategy/irr/strategy.go [hunk 5]
```diff
@@ -398,3 +486,31 @@ func (s *Strategy) CalcAssetValue(price fixedpoint.Value) fixedpoint.Value {
 	balances := s.session.GetAccount().Balances()
 	return balances[s.Market.BaseCurrency].Total().Mul(price).Add(balances[s.Market.QuoteCurrency].Total())
 }
+
+func (s *Strategy) rebalancePosition(bestBid, bestAsk float64, w float64) {
```
Comment (c9s) [inline pkg/strategy/irr/strategy.go:490]: it's better to pass ctx to the function

**Review by c9s — APPROVED** [review]

**Pull request merged**